File: src/invoice-exception/invoice_demo/provider_fixture.py
# SPDX-License-Identifier: Apache-2.0
"""Controlled provider responses prove wiring, not model quality. No oracle access."""

import json
import re
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
from threading import Lock

LOCK = Lock()
STATE = {"mode": "ok", "calls": 0}


class Handler(BaseHTTPRequestHandler):
    def log_message(self, *args):
        pass

    def respond(self, body, status=200):
        raw = json.dumps(body).encode()
        self.send_response(status)
        self.send_header("Content-Type", "application/json")
        self.send_header("Content-Length", str(len(raw)))
        self.end_headers()
        self.wfile.write(raw)

    def do_GET(self):
        if self.path == "/api/tags":
            self.respond({"models": [{"name": "invoice-fixture", "model": "invoice-fixture"}]})
        elif self.path == "/state":
            with LOCK:
                self.respond(dict(STATE))
        else:
            self.respond({"error": "unknown route"}, 404)

    def do_POST(self):
        size = int(self.headers.get("Content-Length", "0"))
        if not 0 < size < 1000000:
            self.respond({"error": "bad request size"}, 400)
            return
        body = json.loads(self.rfile.read(size))
        if self.path == "/mode":
            if body.get("mode") not in ("ok", "unavailable", "bad_citation"):
                self.respond({"error": "unknown mode"}, 400)
                return
            with LOCK:
                STATE["mode"] = body["mode"]
            self.respond({"ok": True})
            return
        if self.path != "/api/chat":
            self.respond({"error": "unknown route"}, 404)
            return
        with LOCK:
            STATE["calls"] += 1
            mode = STATE["mode"]
        if mode == "unavailable":
            self.respond({"error": "synthetic provider outage"}, 503)
            return
        prompt = "\n".join(message["content"] for message in body["messages"])
        facts = json.JSONDecoder().raw_decode(prompt.split("DETERMINISTIC_FACTS=", 1)[1])[0]
        paths = re.findall(r"\[([^\]]+)\] # Fictional purchasing policy", prompt)
        explanation = "Fictional policy review: " + ", ".join(facts["exceptions"] or ["matched"])
        if facts["receipt_status"] == "missing":
            explanation += (
                ". Missing receipt: insufficient evidence of delivery; request the receipt."
            )
        answer = {
            "explanation": explanation,
            "citations": [paths[0] if paths and mode == "ok" else "unserved/policy.md"],
            "recommended_action": "review" if facts["exceptions"] else "acknowledge",
            "receipt_status": facts["receipt_status"],
        }
        self.respond(
            {
                "model": body["model"],
                "done": True,
                "done_reason": "stop",
                "message": {"role": "assistant", "content": json.dumps(answer)},
                "prompt_eval_count": 100,
                "eval_count": 50,
            }
        )


if __name__ == "__main__":
    ThreadingHTTPServer(("0.0.0.0", 11434), Handler).serve_forever()

File: src/quality-investigation/src/main/java/io/ioka/demo/quality/ProviderFixture.java
// SPDX-License-Identifier: Apache-2.0
package io.ioka.demo.quality;
import com.sun.net.httpserver.HttpServer;
import io.ioka.munarium.client.model.Json;
import java.net.*;
import java.nio.charset.StandardCharsets;
import java.util.*;
import java.util.concurrent.atomic.*;
import java.util.regex.Pattern;

public final class ProviderFixture {
    private ProviderFixture() {}
    public static void run() throws Exception {
        var server=HttpServer.create(new InetSocketAddress(11434),0);var mode=new AtomicReference<>("ok");var calls=new AtomicInteger();
        server.createContext("/",exchange->{try {
            int status=200;Object response;String path=exchange.getRequestURI().getPath();
            if(path.equals("/api/tags")) response=Map.of("models",List.of(Map.of("name","quality-fixture","model","quality-fixture")));
            else if(path.equals("/calls")) response=Map.of("calls",calls.get());
            else if(path.startsWith("/mode/")) {mode.set(path.substring(6));response=Map.of("mode",mode.get());}
            else if(path.equals("/api/chat")) {
                calls.incrementAndGet();if(mode.get().equals("unavailable")) {status=503;response=Map.of("error","Synthetic provider outage");}
                else {var body=Json.MAPPER.readTree(exchange.getRequestBody());var prompt=new StringBuilder();for(var message:body.path("messages")) prompt.append(message.path("content").asText()).append('\n');String text=prompt.toString();
                    var fact=Pattern.compile("lot_[0-9]{3}[^\\n]*?defects = ([0-9]+)").matcher(text);var note=Pattern.compile("Narrative defects: ([0-9]+)").matcher(text);var labels=new TreeSet<String>();var matcher=Pattern.compile("\\[([^\\[\\]\\s]+/[^\\[\\]\\s]+)\\]").matcher(text);while(matcher.find()) labels.add(matcher.group(1));
                    var answer=new TreeMap<String,Object>();answer.put("observed_defects",fact.find()?Integer.valueOf(fact.group(1)):null);answer.put("narrative_defects",note.find()?Integer.valueOf(note.group(1)):null);answer.put("action","Hold the lot for quality review.");answer.put("root_cause","not determined");answer.put("hypothesis","requires investigation");answer.put("citations",mode.get().equals("bad-citation")?List.of("unserved/secret"):labels);
                    response=Map.of("model",body.path("model").asText(),"done",true,"done_reason","stop","message",Map.of("role","assistant","content",FilesUtil.json(answer)),"prompt_eval_count",100,"eval_count",90);
                }
            }else {status=404;response=Map.of("error","Unknown route");}
            byte[] bytes=FilesUtil.json(response).getBytes(StandardCharsets.UTF_8);exchange.getResponseHeaders().set("Content-Type","application/json");exchange.sendResponseHeaders(status,bytes.length);exchange.getResponseBody().write(bytes);
        }catch(Exception e) {exchange.sendResponseHeaders(500,-1);}finally {exchange.close();}});server.start();System.out.println("Canned quality protocol fixture; no model inference");
    }
}
